File: web/src/api/resource.ts
import request from './request'
import type { ResourceType, Resource, PaginatedList } from '@/types/resource'

/** 查询资源列表参数 */
interface FetchResourcesParams {
  type: ResourceType
  search?: string
  group_id?: string
  /** 私有资源归属过滤：传 preset_id 仅查该 preset 私有；传 "" 不过滤；不传 → 后端默认只返回全局 */
  owner_preset_id?: string
  page: number
  page_size: number
}

/** 获取资源列表（分页） */
export function fetchResources(params: FetchResourcesParams): Promise<PaginatedList<Resource>> {
  return request.get('/resources', { params })
}

/** 创建资源 */
export function createResource(data: { type: ResourceType; name: string; description?: string }): Promise<Resource> {
  return request.post('/resources', data)
}

/** 获取单个资源详情 */
export function getResource(id: string): Promise<Resource> {
  return request.get(`/resources/${id}`)
}

/** 更新资源信息 */
export function updateResource(id: string, data: { name?: string; description?: string }): Promise<void> {
  return request.put(`/resources/${id}`, data)
}

/** 获取资源内容（解包 content 字段） */
export async function getContent(id: string): Promise<string> {
  const res = await request.get(`/resources/${id}/content`)
  // 后端返回 {content: "..."}, 拦截器解包 data 后是 {content}，再取 .content
  if (res && typeof res === 'object' && 'content' in (res as any)) {
    return (res as any).content
  }
  return res as unknown as string
}

/** 更新资源内容 */
export function updateContent(id: string, content: string): Promise<void> {
  return request.put(`/resources/${id}/content`, { content })
}

/** 删除单个资源
 * @param confirm true=已确认部署提示
 * @param unlink  true=级联解除所有 preset 关联后再删
 */
export function deleteResource(
  id: string,
  options?: { confirm?: boolean; unlink?: boolean },
): Promise<unknown> {
  const params: Record<string, boolean> = {}
  if (options?.confirm) params.confirm = true
  if (options?.unlink) params.unlink = true
  return request.delete(`/resources/${id}`, { params })
}

/** 批量删除资源 */
export function batchDelete(ids: string[], confirm?: boolean): Promise<unknown> {
  return request.delete('/resources/batch', { data: { ids, confirm } })
}

/** 导入 skill: 整目录原样上传,后端落到 ~/.aiManager/skills/{uuid}/
 * @param params.name 从 SKILL.md frontmatter 解析的名称
 * @param params.description 从 frontmatter 解析的描述(可选)
 * @param params.groupId 关联分组(可选)
 * @param params.files 该 skill 子目录下所有文件(File 来自 webkitdirectory)
 * @param params.relPaths 与 files 一一对应的相对路径(相对子目录,如 "SKILL.md"、"assets/x.png")
 */
export function importSkill(params: {
  name: string
  description?: string
  groupId?: string
  files: File[]
  relPaths: string[]
}): Promise<Resource> {
  const fd = new FormData()
  fd.append('name', params.name)
  if (params.description) fd.append('description', params.description)
  if (params.groupId) fd.append('group_id', params.groupId)
  for (let i = 0; i < params.files.length; i++) {
    fd.append('files', params.files[i])
    fd.append('paths', params.relPaths[i])
  }
  return request.post('/resources/import-skill', fd, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
}

/** 导入 agent: 单个 .md 文件原样上传,后端落到 ~/.aiManager/agents/{uuid}.md
 * @param params.name 从 frontmatter 解析的名称
 * @param params.description 从 frontmatter 解析的描述(可空)
 * @param params.groupId 关联分组(可选)
 * @param params.file 源 .md 文件
 */
export function importAgent(params: {
  name: string
  description?: string
  groupId?: string
  file: File
}): Promise<Resource> {
  const fd = new FormData()
  fd.append('name', params.name)
  if (params.description) fd.append('description', params.description)
  if (params.groupId) fd.append('group_id', params.groupId)
  fd.append('file', params.file)
  return request.post('/resources/import-agent', fd, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
}

File: internal/handler/resource.go
// Package handler resource.go 提供资源 CRUD 的 HTTP 接口处理
// 包括列表、创建、详情、更新、删除、批量删除、内容读写
package handler

import (
	"io"
	"strconv"

	"github.com/anthropic/airesourcemanager/internal/model"
	"github.com/anthropic/airesourcemanager/internal/service"
	"github.com/gin-gonic/gin"
)

// ResourceHandler 资源 HTTP 处理器
type ResourceHandler struct {
	svc *service.ResourceService
}

// NewResourceHandler 创建资源处理器实例
// 参数 svc: 资源业务服务
// 返回: ResourceHandler 指针
func NewResourceHandler(svc *service.ResourceService) *ResourceHandler {
	return &ResourceHandler{svc: svc}
}

// RegisterResourceRoutes 注册资源相关路由
// 参数 group: gin 路由组（/api/v1）
// 参数 h: 资源处理器
func RegisterResourceRoutes(group *gin.RouterGroup, h *ResourceHandler) {
	r := group.Group("/resources")
	{
		r.GET("", h.handleList)
		r.POST("", h.handleCreate)
		// import-skill / import-agent 必须在 :id 之前注册
		r.POST("/import-skill", h.handleImportSkill)
		r.POST("/import-agent", h.handleImportAgent)
		// batch 路由必须在 :id 之前注册，避免路由冲突
		r.DELETE("/batch", h.handleBatchDelete)
		r.GET("/:id", h.handleGet)
		r.PUT("/:id", h.handleUpdate)
		r.DELETE("/:id", h.handleDelete)
		r.GET("/:id/content", h.handleGetContent)
		r.PUT("/:id/content", h.handleUpdateContent)
	}
}

// handleList 处理资源列表查询
// 查询参数: type, search, group_id, owner_preset_id, page, page_size
// owner_preset_id 默认为 __none__（仅返回全局资源）；显式传 preset id 时返回该 preset 私有资源
func (h *ResourceHandler) handleList(c *gin.Context) {
	resourceType := c.Query("type")
	search := c.Query("search")
	groupID := c.DefaultQuery("group_id", "0")
	// 默认排除私有资源；若需要查询私有资源前端需传 owner_preset_id={presetID}
	ownerPresetID := c.DefaultQuery("owner_preset_id", "__none__")
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	pageSize, _ := strconv.Atoi(c.DefaultQuery("page_size", "20"))

	resp, err := h.svc.ListResources(resourceType, search, groupID, ownerPresetID, page, pageSize)
	if err != nil {
		handleBizError(c, err)
		return
	}

	Success(c, resp)
}

// handleCreate 处理资源创建
// 请求体: CreateResourceReq
func (h *ResourceHandler) handleCreate(c *gin.Context) {
	var req model.CreateResourceReq
	if err := c.ShouldBindJSON(&req); err != nil {
		Error(c, model.ErrParamValidation, "参数校验失败: "+err.Error())
		return
	}

	resource, err := h.svc.CreateResource(&req)
	if err != nil {
		handleBizError(c, err)
		return
	}

	Success(c, resource)
}

// handleGet 处理资源详情查询
// 路径参数: id
func (h *ResourceHandler) handleGet(c *gin.Context) {
	id := c.Param("id")
	resource, err := h.svc.GetResource(id)
	if err != nil {
		handleBizError(c, err)
		return
	}

	Success(c, resource)
}

// handleUpdate 处理资源元数据更新
// 路径参数: id
// 请求体: UpdateResourceReq
func (h *ResourceHandler) handleUpdate(c *gin.Context) {
	id := c.Param("id")
	var req model.UpdateResourceReq
	if err := c.ShouldBindJSON(&req); err != nil {
		Error(c, model.ErrParamValidation, "参数校验失败: "+err.Error())
		return
	}

	resource, err := h.svc.UpdateResource(id, &req)
	if err != nil {
		handleBizError(c, err)
		return
	}

	Success(c, resource)
}

// handleDelete 处理资源删除
// 路径参数: id
// 查询参数: confirm (true 时级联部署删除), unlink (true 时解除 preset 关联并撤销受影响 preset 部署)
func (h *ResourceHandler) handleDelete(c *gin.Context) {
	id := c.Param("id")
	confirm := c.Query("confirm") == "true"
	unlink := c.Query("unlink") == "true"

	data, err := h.svc.DeleteResource(id, confirm, unlink)
	if err != nil {
		if bizErr, ok := err.(*model.BizError); ok {
			ErrorWithData(c, bizErr.Code, bizErr.Msg, data)
			return
		}
		Error(c, model.ErrResourceFileIO, err.Error())
		return
	}

	Success(c, nil)
}

// handleBatchDelete 处理批量删除
// 请求体: BatchDeleteReq {ids: [], confirm: bool}
// 查询参数: unlink=true 时级联解除 preset 关联
func (h *ResourceHandler) handleBatchDelete(c *gin.Context) {
	var req model.BatchDeleteReq
	if err := c.ShouldBindJSON(&req); err != nil {
		Error(c, model.ErrParamValidation, "参数校验失败: "+err.Error())
		return
	}

	if len(req.IDs) == 0 {
		Error(c, model.ErrParamValidation, "ids 不能为空")
		return
	}

	unlink := c.Query("unlink") == "true"
	results, err := h.svc.BatchDelete(&req, unlink)
	if err != nil {
		handleBizError(c, err)
		return
	}

	Success(c, results)
}

// handleGetContent 处理资源文件内容读取
// 路径参数: id
func (h *ResourceHandler) handleGetContent(c *gin.Context) {
	id := c.Param("id")
	content, err := h.svc.GetContent(id)
	if err != nil {
		handleBizError(c, err)
		return
	}

	Success(c, map[string]string{"content": content})
}

// handleUpdateContent 处理资源文件内容更新
// 路径参数: id
// 请求体: UpdateContentReq
func (h *ResourceHandler) handleUpdateContent(c *gin.Context) {
	id := c.Param("id")
	var req model.UpdateContentReq
	if err := c.ShouldBindJSON(&req); err != nil {
		Error(c, model.ErrParamValidation, "参数校验失败: "+err.Error())
		return
	}

	if err := h.svc.UpdateContent(id, req.Content); err != nil {
		handleBizError(c, err)
		return
	}

	Success(c, nil)
}

// handleImportSkill 处理 skill 目录整体导入
// multipart/form-data 字段:
//   name        — 从源 SKILL.md frontmatter 解析的名称(必填)
//   description — 从源 SKILL.md frontmatter 解析的描述(可选)
//   group_id    — 关联分组 ID(可选)
//   paths       — 文件相对路径数组(与 files 同序,逐个对应)
//   files       — 文件数组
func (h *ResourceHandler) handleImportSkill(c *gin.Context) {
	name := c.PostForm("name")
	description := c.PostForm("description")
	groupID := c.PostForm("group_id")
	paths := c.PostFormArray("paths")

	form, err := c.MultipartForm()
	if err != nil {
		Error(c, model.ErrParamValidation, "解析 multipart 失败: "+err.Error())
		return
	}
	files := form.File["files"]

	if len(paths) != len(files) {
		Error(c, model.ErrParamValidation, "paths 与 files 长度不一致")
		return
	}
	if len(files) == 0 {
		Error(c, model.ErrParamValidation, "至少需要一个文件")
		return
	}

	imported := make([]service.ImportedSkillFile, 0, len(files))
	for i, fh := range files {
		f, err := fh.Open()
		if err != nil {
			Error(c, model.ErrResourceFileIO, "读取上传文件失败: "+err.Error())
			return
		}
		data, err := io.ReadAll(f)
		_ = f.Close()
		if err != nil {
			Error(c, model.ErrResourceFileIO, "读取上传内容失败: "+err.Error())
			return
		}
		imported = append(imported, service.ImportedSkillFile{
			RelPath: paths[i],
			Data:    data,
		})
	}

	resource, err := h.svc.ImportSkill(name, description, groupID, imported)
	if err != nil {
		handleBizError(c, err)
		return
	}
	Success(c, resource)
}

// handleImportAgent 处理 agent 单文件导入
// multipart/form-data 字段:
//   name        — 从源 .md frontmatter 解析的名称(必填)
//   description — 从源 .md frontmatter 解析的描述(可选)
//   group_id    — 关联分组 ID(可选)
//   file        — 源 .md 文件(单文件)
func (h *ResourceHandler) handleImportAgent(c *gin.Context) {
	name := c.PostForm("name")
	description := c.PostForm("description")
	groupID := c.PostForm("group_id")

	fh, err := c.FormFile("file")
	if err != nil {
		Error(c, model.ErrParamValidation, "缺少 file 字段: "+err.Error())
		return
	}
	f, err := fh.Open()
	if err != nil {
		Error(c, model.ErrResourceFileIO, "读取上传文件失败: "+err.Error())
		return
	}
	data, err := io.ReadAll(f)
	_ = f.Close()
	if err != nil {
		Error(c, model.ErrResourceFileIO, "读取上传内容失败: "+err.Error())
		return
	}

	resource, err := h.svc.ImportAgent(name, description, groupID, data)
	if err != nil {
		handleBizError(c, err)
		return
	}
	Success(c, resource)
}

// handleBizError 统一处理业务错误响应
// 参数 c: gin 上下文
// 参数 err: 错误
func handleBizError(c *gin.Context, err error) {
	if bizErr, ok := err.(*model.BizError); ok {
		if bizErr.Data != nil {
			ErrorWithData(c, bizErr.Code, bizErr.Msg, bizErr.Data)
			return
		}
		Error(c, bizErr.Code, bizErr.Msg)
		return
	}
	Error(c, model.ErrResourceFileIO, err.Error())
}
